staging/src/k8s.io/client-go/discovery/discovery_client.go

```diff
@@ -156,12 +174,63 @@ func apiVersionsToAPIGroup(apiVersions *metav1.APIVersions) (apiGroup metav1.API
 	return
 }
 
-// ServerGroups returns the supported groups, with information like supported versions and the
-// preferred version.
-func (d *DiscoveryClient) ServerGroups() (apiGroupList *metav1.APIGroupList, err error) {
+// Returns the discovery groups, and (if new aggregated discovery format) the resources
+// keyed by group/version.
+func (d *DiscoveryClient) GroupsAndMaybeResources() (
+	*metav1.APIGroupList, map[schema.GroupVersion]*metav1.APIResourceList, error) {
+	// First, get the discovery groups/resources at /apis using accept/content-type
+	// negotiation. The accept header enumerates in priority order for the acceptable
+	// discovery content-types. The resultant content-type details the discovery
+	// format returned by the server.
+	var responseContentType string
+	body, err := d.restClient.Get().
+		AbsPath("/apis").
+		SetHeader("Accept", acceptDiscoveryFormats).
+		Do(context.TODO()).
+		ContentType(&responseContentType).
+		Raw()
+	if err != nil && !errors.IsNotFound(err) && !errors.IsForbidden(err) {
+		return nil, nil, err
+	}
+	// Cast the response body into the correct struct base on
+	// discovery response content-type.
+	var apiGroupList *metav1.APIGroupList
+	var resourcesByGV map[schema.GroupVersion]*metav1.APIResourceList
+	if responseContentType == acceptV1 {
```

Comment:
I know there's only two valid options for now, but I'd use a switch for clarity.

Reply:
Fixed.